### PR TITLE
Split trace file into a directory with lazily loaded blocks

### PR DESCRIPTION
**Pull request by alexmojaki**

### CODE REVIEW

File: ice/trace.py [hunk 1]
```diff
@@ -194,3 +261,56 @@ def __new__(mcls, name, bases, namespace):
...
+    result = [get_short_string(v) for v in result]
+    return result
+
+
+def get_short_string(string, max_length=35) -> str:
+    return string[:max_length].strip() + "..." if len(string) > max_length else string
+
+
+def get_short_list(lst: list, max_length=3) -> list:
+    return lst[:max_length] + ["..."] if len(lst) > max_length else lst
```
Comment (lslunis) [inline ice/trace.py:308]: `get_strings` and its support code is kind of messy and hard to follow. (I did a sloppy job of code-reviewing the TracePage versions.) Can you add a TODO to simplify and restructure them?

File: ice/trace.py [hunk 2]
```diff
@@ -194,3 +261,56 @@ def __new__(mcls, name, bases, namespace):
...
+    result = get_short_list(result)
+    result = [get_short_string(v) for v in result]
+    return result
+
+
+def get_short_string(string, max_length=35) -> str:
+    return string[:max_length].strip() + "..." if len(string) > max_length else string
+
+
+def get_short_list(lst: list, max_length=3) -> list:
```
Comment (lslunis) [inline ice/trace.py:307]: Can some of these functions be private?

Comment (lslunis) [inline ice/trace.py:307]: What do you think about tightening up the type annotations for the code in this file, at least for the stuff being added? eg adding missing annotations, and saying things like `list[object]` instead of `list` so we don't end up with implicit anys everywhere.

**Review by lslunis — COMMENTED** [review]
Python comments, now looking at TracePage

File: ui/components/TracePage/TracePage.tsx [hunk 3]
```diff
@@ -207,6 +283,10 @@ const useTreeContext = () => {
   return context;
 };
 
+function useBlockValue<T>(blockAddress: BlockAddress<T>): T | undefined {
```
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:286]: Some of the uses of this hook violate the [rules of hooks](https://reactjs.org/docs/hooks-rules.html). I think it might need to be restructured so a single hook invocation (at the top level) can handle a variable number of block addresses. Want to chat about this?

File: ui/components/TracePage/TracePage.tsx [hunk 4]
```diff
@@ -180,11 +209,58 @@ const TreeProvider = ({ traceId, children }: { traceId: string; children: ReactN
...
+
+    if (blockRequests.current.has(blockNumber)) {
+      // This block has already been requested.
+      return undefined;
+    }
+    blockRequests.current.add(blockNumber);
+
+    const url = `${urlPrefix(traceId)}/block_${blockNumber}.jsonl`;
+    const fetchBlock = async (start: number) => {
+      const response = await fetch(url, { headers: { Range: `bytes=${start}-999999999` } });
```
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:234]: You can leave off the 999999999 and it's still a valid request (but `routes.traces.get_file` will need to be updated to parse that correctly)

File: ui/components/TracePage/TracePage.tsx [hunk 5]
```diff
@@ -180,11 +209,58 @@ const TreeProvider = ({ traceId, children }: { traceId: string; children: ReactN
...
+      const text = await response.text();
+
+      if (!isMounted.current) return;
+
+      const lines = text.split("\n").filter(Boolean);
```
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:239]: Is the filter actually necessary?

File: ui/components/TracePage/TracePage.tsx [hunk 6]
```diff
@@ -180,11 +209,58 @@ const TreeProvider = ({ traceId, children }: { traceId: string; children: ReactN
...
+      if (lines[lines.length - 1] === "end") {
```
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:240]: optional: use lodash's `last`

File: ui/components/TracePage/TracePage.tsx (hunk 5 above)
Comment (lslunis) [inline ui/components/TracePage/TracePage.tsx:239]: I suspect it's possible for the block to be written to disk across multiple system calls, in which case we could get an unterminated fragment of a line that shouldn't be fed into setBlocks

File: ice/trace.py [hunk 7]
```diff
@@ -43,21 +96,24 @@ def _url_prefix():
 
 
 def enable_trace():
-    global trace_file
-
-    trace_file = (trace_dir / f"{trace_id}.jsonl").open("a")
-
-    print(f"Trace: {_url_prefix()}/traces/{trace_id}")
+    trace_var.set(Trace())
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply ice/trace.py:99]: This caused test_all_primer_recipes to fail until I added back a default to parent_id_var. This area is a bit of a mess and should be revisited after this PR. I don't think I can easily raise if a trace is enabled more than once.

File: ice/trace.py (hunk 2 above)
Comment (alexmojaki) [reply ice/trace.py]: `list[object]` doesn't seem any better than `list`.

Comment (lslunis) [reply ice/trace.py]: My understanding is that list is equivalent to list[Any], which disables type-checking of the items. Preferring object to Any is similar to preferring unknown to any in TypeScript.

File: ice/trace.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lslunis) [reply ice/trace.py]: parent_id_var should definitely never be "", so given that Trace is a singleton right now anyway, let's just get rid of self.id for now and change this back to what it was before:

```
trace_id = make_id()
parent_id_var: ContextVar[str] = ContextVar("id", default=trace_id)
```

**Review by lslunis — APPROVED** [review]
